File: src/lib/gamification-constants.ts
// src/lib/gamification-constants.ts
// Sistema de Gamificação Global - Constantes e Traduções

export const RARITY_LABELS: Record<string, string> = {
  common: 'Comum',
  rare: 'Raro',
  epic: 'Épico',
  legendary: 'Lendário',
  mythic: 'Mítico',
};

export const RARITY_COLORS: Record<string, { bg: string; text: string; border: string; gradient: string }> = {
  common: {
    bg: 'rgba(156, 163, 175, 0.1)',
    text: '#6B7280',
    border: 'rgba(156, 163, 175, 0.3)',
    gradient: 'linear-gradient(135deg, #9CA3AF 0%, #6B7280 100%)',
  },
  rare: {
    bg: 'rgba(59, 130, 246, 0.1)',
    text: '#3B82F6',
    border: 'rgba(59, 130, 246, 0.3)',
    gradient: 'linear-gradient(135deg, #60A5FA 0%, #3B82F6 100%)',
  },
  epic: {
    bg: 'rgba(139, 92, 246, 0.1)',
    text: '#8B5CF6',
    border: 'rgba(139, 92, 246, 0.3)',
    gradient: 'linear-gradient(135deg, #A78BFA 0%, #8B5CF6 100%)',
  },
  legendary: {
    bg: 'rgba(245, 158, 11, 0.1)',
    text: '#F59E0B',
    border: 'rgba(245, 158, 11, 0.3)',
    gradient: 'linear-gradient(135deg, #FBBF24 0%, #F59E0B 100%)',
  },
  mythic: {
    bg: 'rgba(236, 72, 153, 0.1)',
    text: '#EC4899',
    border: 'rgba(236, 72, 153, 0.3)',
    gradient: 'linear-gradient(135deg, #F472B6 0%, #EC4899 100%)',
  },
};

export const LEVEL_NAMES: Record<number, { name: string; title: string; icon: string }> = {
  1: { name: 'Iniciante', title: 'Aprendiz Financeiro', icon: '🌱' },
  2: { name: 'Organizador', title: 'Controlador de Gastos', icon: '📊' },
  3: { name: 'Disciplinado', title: 'Guardião do Orçamento', icon: '🎯' },
  4: { name: 'Estrategista', title: 'Mestre do Planejamento', icon: '🧠' },
  5: { name: 'Expert', title: 'Sábio das Finanças', icon: '⚡' },
  6: { name: 'Veterano', title: 'Guru Financeiro', icon: '🏆' },
  7: { name: 'Elite', title: 'Lenda Econômica', icon: '💎' },
  8: { name: 'Mestre', title: 'Senhor das Finanças', icon: '👑' },
  9: { name: 'Grão-Mestre', title: 'Imperador Financeiro', icon: '🌟' },
  10: { name: 'Lenda', title: 'Transcendente', icon: '✨' },
};

// Pontos por nível (progressão exponencial)
export const LEVEL_THRESHOLDS = [0, 100, 300, 600, 1000, 1500, 2200, 3000, 4000, 5500];

// XP por ação
export const XP_REWARDS = {
  // Transações
  ADD_TRANSACTION: 5,
  CATEGORIZE_TRANSACTION: 2,
  FIRST_TRANSACTION_TODAY: 10,

  // Parcelamentos
  PAY_INSTALLMENT: 10,
  PAY_ON_TIME_BONUS: 5,
  COMPLETE_INSTALLMENT: 50,
  LATE_PAYMENT_PENALTY: -2, // por dia

  // Orçamentos
  CREATE_BUDGET: 15,
  STAY_WITHIN_BUDGET: 20,
  PERFECT_BUDGET_MONTH: 100,

  // Metas
  CREATE_GOAL: 10,
  CONTRIBUTE_TO_GOAL: 5,
  COMPLETE_GOAL: 100,

  // Carteiras
  CREATE_WALLET: 10,
  BALANCE_WALLETS: 25,

  // Uso do app
  DAILY_LOGIN: 5,
  WEEKLY_STREAK: 25,
  USE_AI_ASSISTANT: 3,
  IMPORT_TRANSACTIONS: 20,
  VIEW_REPORTS: 5,
  USE_CALCULATOR: 3,

  // Desafios
  COMPLETE_DAILY_QUEST: 15,
  COMPLETE_WEEKLY_CHALLENGE: 50,
  COMPLETE_MONTHLY_CHALLENGE: 150,
};

// Badges disponíveis
export const ALL_BADGES = [
  // Iniciante
  { id: 'first-steps', name: 'Primeiros Passos', description: 'Criou sua primeira transação', icon: '👣', rarity: 'common' as const, category: 'onboarding' },
  { id: 'wallet-creator', name: 'Organizador', description: 'Criou sua primeira carteira', icon: '💼', rarity: 'common' as const, category: 'onboarding' },
  { id: 'budget-starter', name: 'Planejador Iniciante', description: 'Criou seu primeiro orçamento', icon: '📋', rarity: 'common' as const, category: 'onboarding' },
  { id: 'goal-setter', name: 'Sonhador', description: 'Definiu sua primeira meta', icon: '🎯', rarity: 'common' as const, category: 'onboarding' },

  // Consistência
  { id: 'week-streak', name: 'Semana Perfeita', description: '7 dias usando o app', icon: '📅', rarity: 'common' as const, category: 'consistency' },
  { id: 'month-streak', name: 'Mês Dedicado', description: '30 dias usando o app', icon: '🗓️', rarity: 'rare' as const, category: 'consistency' },
  { id: 'quarter-streak', name: 'Trimestre Focado', description: '90 dias usando o app', icon: '📆', rarity: 'epic' as const, category: 'consistency' },
  { id: 'year-streak', name: 'Ano de Compromisso', description: '365 dias usando o app', icon: '🎊', rarity: 'legendary' as const, category: 'consistency' },

  // Pagamentos
  { id: 'first-payment', name: 'Pagador', description: 'Pagou sua primeira parcela', icon: '💳', rarity: 'common' as const, category: 'payments' },
  { id: 'punctual-10', name: 'Pontual', description: '10 pagamentos em dia', icon: '⏰', rarity: 'rare' as const, category: 'payments' },
  { id: 'punctual-50', name: 'Super Pontual', description: '50 pagamentos em dia', icon: '⏱️', rarity: 'epic' as const, category: 'payments' },
  { id: 'punctual-100', name: 'Mestre da Pontualidade', description: '100 pagamentos em dia', icon: '🕐', rarity: 'legendary' as const, category: 'payments' },
  { id: 'zero-delay', name: 'Impecável', description: 'Nunca atrasou um pagamento (mín. 20)', icon: '✨', rarity: 'mythic' as const, category: 'payments' },

  // Parcelamentos
  { id: 'installment-complete-1', name: 'Finalizador', description: 'Completou 1 parcelamento', icon: '🏁', rarity: 'common' as const, category: 'installments' },
  { id: 'installment-complete-5', name: 'Quitador', description: 'Completou 5 parcelamentos', icon: '🎖️', rarity: 'rare' as const, category: 'installments' },
  { id: 'installment-complete-15', name: 'Livre de Dívidas', description: 'Completou 15 parcelamentos', icon: '🏆', rarity: 'epic' as const, category: 'installments' },
  { id: 'installment-complete-30', name: 'Destruidor de Dívidas', description: 'Completou 30 parcelamentos', icon: '💪', rarity: 'legendary' as const, category: 'installments' },

  // Orçamentos
  { id: 'budget-success-1', name: 'Controlado', description: 'Ficou dentro do orçamento 1 mês', icon: '📊', rarity: 'common' as const, category: 'budgets' },
  { id: 'budget-success-3', name: 'Disciplinado', description: 'Ficou dentro do orçamento 3 meses', icon: '📈', rarity: 'rare' as const, category: 'budgets' },
  { id: 'budget-success-6', name: 'Mestre do Controle', description: 'Ficou dentro do orçamento 6 meses', icon: '🎯', rarity: 'epic' as const, category: 'budgets' },
  { id: 'budget-success-12', name: 'Rei do Orçamento', description: 'Ficou dentro do orçamento 12 meses', icon: '👑', rarity: 'legendary' as const, category: 'budgets' },

  // Metas
  { id: 'goal-complete-1', name: 'Realizador', description: 'Completou 1 meta', icon: '🌟', rarity: 'common' as const, category: 'goals' },
  { id: 'goal-complete-5', name: 'Conquistador', description: 'Completou 5 metas', icon: '⭐', rarity: 'rare' as const, category: 'goals' },
  { id: 'goal-complete-10', name: 'Campeão de Metas', description: 'Completou 10 metas', icon: '🏅', rarity: 'epic' as const, category: 'goals' },
  { id: 'goal-1000', name: 'Poupador Bronze', description: 'Economizou R$ 1.000', icon: '🥉', rarity: 'common' as const, category: 'goals' },
  { id: 'goal-5000', name: 'Poupador Prata', description: 'Economizou R$ 5.000', icon: '🥈', rarity: 'rare' as const, category: 'goals' },
  { id: 'goal-10000', name: 'Poupador Ouro', description: 'Economizou R$ 10.000', icon: '🥇', rarity: 'epic' as const, category: 'goals' },
  { id: 'goal-50000', name: 'Poupador Diamante', description: 'Economizou R$ 50.000', icon: '💎', rarity: 'legendary' as const, category: 'goals' },

  // Economia
  { id: 'saver-month', name: 'Econômico', description: 'Gastou menos que o mês anterior', icon: '📉', rarity: 'common' as const, category: 'savings' },
  { id: 'saver-10percent', name: 'Poupador 10%', description: 'Economizou 10% da renda', icon: '💰', rarity: 'rare' as const, category: 'savings' },
  { id: 'saver-20percent', name: 'Poupador 20%', description: 'Economizou 20% da renda', icon: '💵', rarity: 'epic' as const, category: 'savings' },
  { id: 'saver-30percent', name: 'Investidor Nato', description: 'Economizou 30% da renda', icon: '🤑', rarity: 'legendary' as const, category: 'savings' },

  // Especiais
  { id: 'no-spend-day', name: 'Dia Econômico', description: 'Passou um dia sem gastar', icon: '🙅', rarity: 'common' as const, category: 'special' },
  { id: 'no-spend-week', name: 'Semana Frugal', description: 'Passou uma semana gastando pouco', icon: '🧘', rarity: 'rare' as const, category: 'special' },
  { id: 'early-bird', name: 'Madrugador', description: 'Usou o app antes das 6h', icon: '🌅', rarity: 'common' as const, category: 'special' },
  { id: 'night-owl', name: 'Coruja', description: 'Usou o app depois da meia-noite', icon: '🦉', rarity: 'common' as const, category: 'special' },
  { id: 'ai-friend', name: 'Amigo da IA', description: 'Usou o assistente de IA 50 vezes', icon: '🤖', rarity: 'rare' as const, category: 'special' },
  { id: 'tool-master', name: 'Calculista', description: 'Usou todas as calculadoras', icon: '🧮', rarity: 'rare' as const, category: 'special' },
  { id: 'report-lover', name: 'Analista', description: 'Gerou 10 relatórios', icon: '📑', rarity: 'rare' as const, category: 'special' },

  // Recuperação
  { id: 'comeback', name: 'Volta por Cima', description: 'Quitou parcelas em atraso', icon: '💪', rarity: 'rare' as const, category: 'recovery' },
  { id: 'debt-free', name: 'Livre!', description: 'Zerou todas as dívidas', icon: '🎉', rarity: 'epic' as const, category: 'recovery' },
];

// Desafios diários possíveis
export const DAILY_QUESTS = [
  { id: 'add-transaction', name: 'Registrar Hoje', description: 'Adicione uma transação hoje', xp: 15, icon: '📝' },
  { id: 'check-budgets', name: 'Revisar Orçamentos', description: 'Visualize seus orçamentos', xp: 10, icon: '📊' },
  { id: 'check-goals', name: 'Acompanhar Metas', description: 'Visualize suas metas', xp: 10, icon: '🎯' },
  { id: 'use-ai', name: 'Consultar IA', description: 'Faça uma pergunta ao assistente', xp: 10, icon: '🤖' },
  { id: 'categorize', name: 'Organizar', description: 'Categorize uma transação', xp: 10, icon: '🏷️' },
  { id: 'view-report', name: 'Analisar', description: 'Visualize um relatório', xp: 10, icon: '📈' },
];

// Desafios semanais possíveis
export const WEEKLY_CHALLENGES = [
  { id: 'budget-week', name: 'Semana no Orçamento', description: 'Fique dentro do orçamento por 7 dias', xp: 50, icon: '📋' },
  { id: 'save-week', name: 'Semana Econômica', description: 'Gaste 10% menos que a semana passada', xp: 75, icon: '💰' },
  { id: 'no-unnecessary', name: 'Essencial', description: 'Evite gastos supérfluos por 5 dias', xp: 60, icon: '🎯' },
  { id: 'register-all', name: 'Registrador', description: 'Registre todas as transações da semana', xp: 40, icon: '📝' },
  { id: 'goal-contribution', name: 'Investidor', description: 'Contribua para uma meta 3x na semana', xp: 50, icon: '🎯' },
];

// Desafios mensais possíveis
export const MONTHLY_CHALLENGES = [
  { id: 'perfect-month', name: 'Mês Perfeito', description: 'Complete todos os desafios diários do mês', xp: 200, icon: '🏆' },
  { id: 'budget-master', name: 'Mestre do Orçamento', description: 'Fique dentro do orçamento o mês inteiro', xp: 150, icon: '👑' },
  { id: 'save-goal', name: 'Meta Alcançada', description: 'Complete uma meta de economia', xp: 150, icon: '🎯' },
  { id: 'debt-reduction', name: 'Redutor de Dívidas', description: 'Pague todas as parcelas do mês em dia', xp: 100, icon: '💳' },
  { id: 'categorize-all', name: 'Organizador Total', description: 'Categorize 100% das transações', xp: 100, icon: '🏷️' },
];

// Função para obter label de raridade em português
export function getRarityLabel(rarity: string): string {
  return RARITY_LABELS[rarity] || rarity;
}

// Função para obter cores de raridade
export function getRarityColors(rarity: string) {
  return RARITY_COLORS[rarity] || RARITY_COLORS.common;
}

// Função para obter info do nível
export function getLevelInfo(level: number) {
  return LEVEL_NAMES[Math.min(level, 10)] || LEVEL_NAMES[10];
}

// Função para calcular nível baseado em XP
export function calculateLevel(xp: number): { level: number; currentXp: number; xpForNext: number; progress: number } {
  let level = 1;
  for (let i = LEVEL_THRESHOLDS.length - 1; i >= 0; i--) {
    if (xp >= LEVEL_THRESHOLDS[i]) {
      level = i + 1;
      break;
    }
  }

  const currentThreshold = LEVEL_THRESHOLDS[level - 1] || 0;
  const nextThreshold = LEVEL_THRESHOLDS[level] || LEVEL_THRESHOLDS[LEVEL_THRESHOLDS.length - 1] + 1000;
  const xpInLevel = xp - currentThreshold;
  const xpNeeded = nextThreshold - currentThreshold;

  return {
    level,
    currentXp: xpInLevel,
    xpForNext: xpNeeded,
    progress: Math.min((xpInLevel / xpNeeded) * 100, 100),
  };
}

File: mobile/lib/core/constants/gamification_constants.dart
import 'package:flutter/material.dart';

// Constantes de Gamificação portadas do Web

class GamificationConstants {
  static const Map<int, LevelInfo> levelNames = {
    1: LevelInfo('Iniciante', 'Aprendiz Financeiro', '🌱'),
    2: LevelInfo('Organizador', 'Controlador de Gastos', '📊'),
    3: LevelInfo('Disciplinado', 'Guardião do Orçamento', '🎯'),
    4: LevelInfo('Estrategista', 'Mestre do Planejamento', '🧠'),
    5: LevelInfo('Expert', 'Sábio das Finanças', '⚡'),
    6: LevelInfo('Veterano', 'Guru Financeiro', '🏆'),
    7: LevelInfo('Elite', 'Lenda Econômica', '💎'),
    8: LevelInfo('Mestre', 'Senhor das Finanças', '👑'),
    9: LevelInfo('Grão-Mestre', 'Imperador Financeiro', '🌟'),
    10: LevelInfo('Lenda', 'Transcendente', '✨'),
  };

  static const List<int> levelThresholds = [0, 100, 300, 600, 1000, 1500, 2200, 3000, 4000, 5500];

  static const Map<String, int> xpRewards = {
    // Transações
    'Nova Transação': 5,
    'Categorizar': 2,
    '1ª do Dia': 10,
    // Parcelamentos
    'Pagar Parcela': 10,
    'Pagamento em Dia': 5,
    'Completar Carnê': 50,
    // Orçamentos
    'Criar Orçamento': 15,
    'Dentro da Meta': 20,
    'Mês Perfeito': 100,
    // Metas
    'Criar Meta': 10,
    'Contribuir': 5,
    'Concluir Meta': 100,
    // Uso
    'Login Diário': 5,
    'Streak Semanal': 25,
    'Usar IA': 3,
  };

  static const List<BadgeInfo> allBadges = [
    // Iniciante
    BadgeInfo('first-steps', 'Primeiros Passos', 'Criou sua primeira transação', '👣', 'common', 'onboarding'),
    BadgeInfo('wallet-creator', 'Organizador', 'Criou sua primeira carteira', '💼', 'common', 'onboarding'),
    BadgeInfo('goal-setter', 'Sonhador', 'Definiu sua primeira meta', '🎯', 'common', 'onboarding'),
    
    // Consistência
    BadgeInfo('week-streak', 'Semana Perfeita', '7 dias usando o app', '📅', 'common', 'consistency'),
    BadgeInfo('month-streak', 'Mês Dedicado', '30 dias usando o app', '🗓️', 'rare', 'consistency'),
    BadgeInfo('year-streak', 'Lenda Viva', '365 dias usando o app', '🎊', 'legendary', 'consistency'),

    // Pagamentos
    BadgeInfo('first-payment', 'Pagador', 'Pagou sua primeira parcela', '💳', 'common', 'payments'),
    BadgeInfo('punctual-10', 'Pontual', '10 pagamentos em dia', '⏰', 'rare', 'payments'),
    BadgeInfo('zero-delay', 'Impecável', 'Nunca atrasou (mín. 20)', '✨', 'mythic', 'payments'),

    // Economia
    BadgeInfo('saver-month', 'Econômico', 'Gastou menos que mês anterior', '📉', 'common', 'savings'),
    BadgeInfo('saver-20percent', 'Poupador 20%', 'Economizou 20% da renda', '💵', 'epic', 'savings'),
    
    // Especiais
    BadgeInfo('early-bird', 'Madrugador', 'Usou antes das 6h', '🌅', 'common', 'special'),
    BadgeInfo('night-owl', 'Coruja', 'Usou após meia-noite', '🦉', 'common', 'special'),
    BadgeInfo('ai-friend', 'Amigo da IA', 'Usou IA 50 vezes', '🤖', 'rare', 'special'),
  ];

  static const List<QuestInfo> dailyQuests = [
    QuestInfo('add-transaction', 'Registrar Hoje', 'Adicione uma transação hoje', '📝', 15),
    QuestInfo('check-budgets', 'Revisar Orçamentos', 'Visualize seus orçamentos', '📊', 10),
    QuestInfo('check-goals', 'Acompanhar Metas', 'Visualize suas metas', '🎯', 10),
    QuestInfo('use-ai', 'Consultar IA', 'Faça uma pergunta ao assistente', '🤖', 10),
    QuestInfo('categorize', 'Organizar', 'Categorize uma transação', '🏷️', 10),
    QuestInfo('view-report', 'Analisar', 'Visualize um relatório', '📈', 10),
  ];

  static const List<QuestInfo> weeklyChallenges = [
    QuestInfo('budget-week', 'Semana no Orçamento', 'Fique dentro do orçamento por 7 dias', '📋', 50),
    QuestInfo('save-week', 'Semana Econômica', 'Gaste 10% menos que a semana passada', '💰', 75),
    QuestInfo('no-unnecessary', 'Essencial', 'Evite gastos supérfluos por 5 dias', '🎯', 60),
    QuestInfo('register-all', 'Registrador', 'Registre todas as transações da semana', '📝', 40),
  ];

  static const List<QuestInfo> monthlyChallenges = [
    QuestInfo('perfect-month', 'Mês Perfeito', 'Complete todos as missões diárias', '🏆', 200),
    QuestInfo('budget-master', 'Mestre do Orçamento', 'Fique dentro do orçamento o mês todo', '👑', 150),
    QuestInfo('save-goal', 'Meta Alcançada', 'Complete uma meta de economia', '🎯', 150),
    QuestInfo('debt-reduction', 'Redutor de Dívidas', 'Pague todas as parcelas em dia', '💳', 100),
  ];

  static const Map<String, RarityColor> rarityColors = {
    'common': RarityColor(Color(0xFF6B7280), Color(0x1A9CA3AF), Color(0xFF9CA3AF)),
    'rare': RarityColor(Color(0xFF3B82F6), Color(0x1A3B82F6), Color(0xFF3B82F6)),
    'epic': RarityColor(Color(0xFF8B5CF6), Color(0x1A8B5CF6), Color(0xFF8B5CF6)),
    'legendary': RarityColor(Color(0xFFF59E0B), Color(0x1AF59E0B), Color(0xFFF59E0B)),
    'mythic': RarityColor(Color(0xFFEC4899), Color(0x1AEC4899), Color(0xFFEC4899)),
  };

  static String getRarityLabel(String rarity) {
    switch (rarity) {
      case 'common': return 'Comum';
      case 'rare': return 'Raro';
      case 'epic': return 'Épico';
      case 'legendary': return 'Lendário';
      case 'mythic': return 'Mítico';
      default: return rarity;
    }
  }
}

class LevelInfo {
  final String name;
  final String title;
  final String icon;
  const LevelInfo(this.name, this.title, this.icon);
}

class BadgeInfo {
  final String id;
  final String name;
  final String description;
  final String icon;
  final String rarity;
  final String category;
  const BadgeInfo(this.id, this.name, this.description, this.icon, this.rarity, this.category);
}

class RarityColor {
  final Color text;
  final Color bg;
  final Color border;
  const RarityColor(this.text, this.bg, this.border);
}

class QuestInfo {
  final String id;
  final String name;
  final String description;
  final String icon;
  final int xp;
  const QuestInfo(this.id, this.name, this.description, this.icon, this.xp);
}
